lib/http.c

```diff
@@ -696,6 +696,15 @@ CURLcode Curl_http_auth_act(struct Curl_easy *data)
     data->req.newurl = strdup(data->state.url); /* clone URL */
     if(!data->req.newurl)
       return CURLE_OUT_OF_MEMORY;
+#ifndef CURL_DISABLE_RTSP
+    /*
+     * Authentication is treated as a redirect in Curl_follow(), so if this is
+     * done using RTSP we make it allow these "redirects" to RTSP (only). A
+     * safe assumption as no other redirects should happen from RTSP.
+     */
+    if(conn->handler->protocol & CURLPROTO_RTSP)
+      data->set.redir_protocols = CURLPROTO_RTSP;
```

Comment:
This probably breaks reuse of handle with another protocol.
If handle is used for RTSP then for HTTP, redirection won't work on HTTP unless reset externally between both uses.
I would rather use a copy of `set.redir_protocols` into a new `state.redir_protocol` field in the whole library.

Reply:
Yikes, that is indeed a good remark. I need to fix this...

Reply:
The RTSP code also wrongly sets `data->set.opt_no_body` in several places...

Reply:
I did not see ithem before but you're right. I noticed this particular one while rebasing the sieve PR, as it also features redirection.

I've just done a quick check for more and there may be some (although not all listed):

```
$ find lib -name '*.c' | xargs grep -e '->set\.[A-Za-z0-9_.]* *= ' | grep -Fv 'setopt.c
> url.c'
lib/easy.c:    dst->set.postfields = dst->set.str[i];
lib/easy.c:  outcurl->set.buffer_size = data->set.buffer_size;
lib/http.c:      data->set.redir_protocols = CURLPROTO_RTSP;
lib/ftp.c:    data->set.ftp_filemethod = FTPFILE_MULTICWD;
lib/ftp.c:  data->set.fwrite_func = Curl_ftp_parselist;
lib/ftp.c:  data->set.out = data;
lib/ftp.c:      data->set.fwrite_func = ftpwc->backup.write_function;
lib/ftp.c:      data->set.out = ftpwc->backup.file_descriptor;
lib/doh.c:    doh->set.fmultidone = doh_done;
lib/doh.c:    doh->set.dohfor = data; /* identify for which transfer this is done */
lib/rtsp.c:  data->set.opt_no_body = TRUE; /* most requests don't contain a body */
lib/rtsp.c:    data->set.opt_no_body = FALSE;
lib/rtsp.c:    data->set.opt_no_body = FALSE;
lib/rtsp.c:    data->set.opt_no_body = FALSE;
lib/rtsp.c:      data->set.opt_no_body = TRUE;
lib/transfer.c:      data->set.trailer_data = NULL;
lib/transfer.c:      data->set.trailer_callback = NULL;
lib/transfer.c:      data->set.upload = false;
lib/http2.c:        node->data->set.stream_depends_on = child;
lib/http2.c:      parent->set.stream_dependents = 0;
lib/http2.c:      (*tail)->data->set.stream_depends_e = FALSE;
lib/http2.c:  child->set.stream_depends_on = parent;
lib/http2.c:  child->set.stream_depends_e = exclusive;
lib/http2.c:      parent->set.stream_dependents = data->next;
lib/http2.c:  child->set.stream_depends_on = 0;
lib/http2.c:  child->set.stream_depends_e = FALSE;
lib/multi.c:  data->state.conn_cache->closure_handle->set.timeout = data->set.timeout;
lib/conncache.c:  connc->closure_handle->set.buffer_size = READBUFFER_MIN;
lib/vtls/vtls.c:  data->set.general_ssl.max_ssl_sessions = amount;
$
```
Maybe we can plan to have only setopt values in `set` and r/w runtime values in `state`? Just an idea...

Reply:
> Maybe we can plan to have only setopt values in set and r/w runtime values in state?

I think we should, yes. And then possibly invent and set up some some kind of framework/markup or something that lets us scan for mistakes like the present.

Reply:
#9888 addresses parts of this

Reply:
> And then possibly invent and set up some some kind of framework/markup

Suggestion:

- declare `set` as `const`
- use a cast with a comment in the few function allowed to alter it.

Not very elegant but will cause a compile error if such a mistake occurs again. The C language does not offer many alternatives to achieve that. And it does not require any external test or tool.